### PR TITLE
[Enhancement] auto detect and set jvm xmx parameter

### LINKED ISSUE #31521: [FE][JVM] FE jvm parameter auto detect and setting in container env
## Enhancement

Be able to auto detect mem limit in container env and properly set  JVM `-Xmx` parameter.

Right now if the user forgets to adjust the -Xmx in fe.conf, even the container is assigned with a large number of memory, the FE will still use only -Xmx number of memory without even notice.

Proposed Solution:

Add two additional environment in FE start script.
1. FE_ENABLE_AUTO_JVM_XMX_DETECT=true|false, default to false, if enabled, the start script will check the memory limit and assign a proper Xmx configuration.
2. FE_JVM_XMX_PERCENTAGE=##, the number of percentage to be assigned to -Xmx. default value: 70, user can set this env to a larger or smaller value as necessary.

So the startup script will do this.

1. check if $FE_ENABLE_AUTO_JVM_XMX_DETECT=true
2. if $FE_ENABLE_AUTO_JVM_XMX_DETECT=true, auto detect the memory limit MEM_LIMIT of the container
3. get $FE_JVM_XMX_PERCENTAGE, if not set, assume it is 70 (70%)
4. set MX=$MEM_LIMIT * $FE_JVM_XMX_PERCENTAGE
5. update JVM_OPTS, set `-Xmx${MX}`

### PR DESCRIPTION
**Pull request by yandongxiao**
Fixes #31521 

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by yandongxiao** [comment]
my test:

values.yaml
```
operator:
  starrocksOperator:
    image:
      repository: starrocks/operator
      tag: v1.8.3
    imagePullPolicy: IfNotPresent
  timeZone: Asia/Shanghai

starrocks:
  starrocksBeSpec:
    image:
      repository:  xxx
      tag: 3.1.2
    resources:
      limits:
        cpu: 8
        memory: 8Gi
      requests:
        cpu: 101m
        memory: 100Mi
  starrocksFESpec:
    image:
      repository: xxx
      tag: "20231009_185820"
    feEnvVars:
      - name: FE_ENABLE_AUTO_JVM_XMX_DETECT
        value: "true"
    resources:
      limits:
        cpu: 8
        memory: 8Gi
      requests:
        cpu: 102m
        memory: 100Mi
  timeZone: Asia/Shanghai
```

the JVM parameter output

```
jps -v
400 Jps -Dapplication.home=/usr/lib/jvm/java-11-openjdk-amd64 -Xms8m -Djdk.module.main=jdk.jcmd
134 StarRocksFE -Dlog4j2.formatMsgNoLookups=true -Xmx8192m -XX:+UseG1GC -Xlog:gc*:/opt/starrocks/fe/log/fe.gc.log.20231009-191011:time -Xmx5734m
```

File: bin/common.sh [hunk 1]
```diff
@@ -98,3 +98,41 @@ update_submodules()
...
+  if [[ ! "$MEM_LIMIT" =~ ^[0-9]+$ ]]; then
+    return
+  fi
+
+  if [[ -z "$FE_JVM_XMX_PERCENTAGE" ]]; then
+    FE_JVM_XMX_PERCENTAGE=70
+  fi
+
+ # shellcheck disable=SC1102
+ MX=$(($MEM_LIMIT * $FE_JVM_XMX_PERCENTAGE / 100 / 1024 / 1024))
```
Comment (kevincai) [inline bin/common.sh:136]:
```suggestion
let "MX = MEM_LIMIT * FE_JVM_XMX_PERCENTAGE / 100 / 1024 / 1024"
```

File: bin/start_fe.sh [hunk 2]
```diff
@@ -128,6 +128,11 @@ if [[ "$JAVA_VERSION" -gt 8 ]]; then
     fi
 fi
 
+# detect xmx
+# if detect_jvm_xmx failed to detect xmx, xmx will be empty
+xmx=$(detect_jvm_xmx)
+final_java_opt="${final_java_opt} ${xmx}"
```
[the remark this answers is not in the document]
Comment (yandongxiao) [reply bin/start_fe.sh:134]: the last -Xmx will take effect

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: bin/common.sh [hunk 3]
```diff
@@ -98,3 +98,41 @@ update_submodules()
     git submodule update --init --recursive
     popd
 }
+
+# if $FE_ENABLE_AUTO_JVM_XMX_DETECT=true, auto detect the memory limit MEM_LIMIT of the container
+# get $FE_JVM_XMX_PERCENTAGE, if not set, assume it is 70.
+# the xmx value will be $MEM_LIMIT * $FE_JVM_XMX_PERCENTAGE / 100 / 1024 / 1024
+# output string, e.g. -Xmx4096m
+detect_jvm_xmx() {
```
Comment (kevincai) [inline bin/common.sh:106]: seems code indention is not consistent with other parts

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by kevincai — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Review by sduzh — APPROVED** [review]

**Comment by yandongxiao** [comment]
@mergify backport branch-3.2 branch-3.1 branch-3.0 branch-2.5

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**